### DISCUSSION TITLE
Minion::Backend::Pg: Postpone migrations to first connection

### DISCUSSION THREAD

**Discussion by augensalat** [mojolicious/mojo#1886] [Q&A]
After switching from `Minion::Backend::mysql` to `Minion::Backend::Pg` I have issues with application commands:

The constructor of `Minion::Backend::Pg` tries to run its migrations as soon it is instantiated - this seems to be the case if you have

```perl
$app->plugin(Minion => {Pg => $connection_string});
```

in your application class.

As a consequence migrations are tried even when a simple `Mojolicous::Command` is executed by an application, i.e.

```
my_app version
```

This command fails when the database - which is not required for the command - is not accessible. OTOH the migration is always tried if the database can be reached, which is not always desired, and slows down the commands in any case.

In a project based on Mojolicious we use a couple of dev commands in the `Author` namespace (i.e. for CI/CD) that don't have and don't need access to the database. This fails now.

Currently I'm not aware of a way to find out in the application code by which `Mojolicous::Command` it was started in order to conditionally skip the Minion-plugin loading. Even then it would be a clumsy work-around.

As indicated I did not have these issues with `Minion::Backend::mysql` that postpones migration to the first database connection event. I would really appreciate If `Minion::Backend::Pg` would do that in the same way.

What do you think?

**Comment by kraih** [comment]
This is incorrect, `Minion::Backend::Pg` does nothing with the migration after preparing it in the constructor. https://github.com/mojolicious/minion/blob/f4bc5b2e98aa619b88cd1c6188649f281be1c49e/lib/Minion/Backend/Pg.pm#L135

**Reply by augensalat** [reply]
Isn't the `$self->pg->auto_migrate(1)` exactly what runs the migration immediately?

**Reply by augensalat** [reply]
I also think that the connection already happens in
https://github.com/mojolicious/minion/blob/f4bc5b2e98aa619b88cd1c6188649f281be1c49e/lib/Minion/Backend/Pg.pm#L130

**Reply by augensalat** [reply]
... or line 131.

**Reply by kraih** [reply]
No, the migration is not loaded there. And `auto_migrate` is just a plain old attribute.

**Reply by augensalat** [reply]
OK. Maybe I'm wrong and the migration is not loaded there.
But the constructor connects to the database server to determine the server version. This is still a problem for the commands that do not have/need database access.

**Reply by augensalat** [reply]
Just checked: I was wrong - the migration is not executed by the web app, but when the Minion scheduler is started.
However my main problem remains: The application always tries to connect to the database at `startup`, cause by this check:
https://github.com/mojolicious/minion/blob/f4bc5b2e98aa619b88cd1c6188649f281be1c49e/lib/Minion/Backend/Pg.pm#L130-L132
and at the moment I don't have a (simple) solution how to solve the problem.

**Comment by s1037989** [comment]
Then perhaps they shouldn't initialize Mojo::Pg?

Just as an example, initialize if app->mode eq production, but if mode eq
cli, don't.

Of course mode is not the best condition to base this on, but it
illustrates the point. Environment variable? Config value?  I'd think
there's some way for a command to influence the startup of the application,
but I can't think of the most ideal at the moment.

**Reply by augensalat** [reply]
My Mojolicious application uses `Mojolicious::Plugin::Minion` (see initial question) which, in the startup phase, instantiates a `Minion` object which instantiates a `Minion::Backend::Pg` object which connects to the database, what eventually might fail.

Where would you set the `mode` or any other environment variable? The only way that I can imagine is

```
MOJO_MODE=cli my_app version
```

and 

```perl
$app->plugin(Minion => {Pg => $connection_string}) if $app->mode ne 'cli';
```

but this is ugly, hackish and error-prone at the same time IMHO - and perhaps fails somewhere else because for `mode` only "development" and "production" are documented as valid values.
Same for configs. There is a config for each, development and production. Both environments need Minion when running the web application.

**Reply by s1037989** [reply]
The `-m, --mode <name>` option is [available for all commands](https://docs.mojolicious.org/Mojolicious/Commands), so that will resolve that bit of ugliness.

```bash
$ my_app version -m cli
```

And the solution to your problem is to use the [before_command](https://docs.mojolicious.org/Mojolicious#before_command) hook.
```perl
$ cat myapp.pl 
#!/usr/bin/env perl
package Mojolicious::Command::no_pg;
use Mojo::Base 'Mojolicious::Command';
sub run {say shift->app->pg }

package Mojolicious::Command::with_pg;
use Mojo::Base 'Mojolicious::Command';
sub run { say shift->app->pg }

package main;
use Mojolicious::Lite -signatures;
hook before_command => sub ($command, $args) {
  $command->app->helper(pg => ref($command) =~ /no_pg$/ ? sub {'no_pg'} : sub {'with_pg'});
};
app->start;
```
Results:
```bash
$ perl myapp.pl with_pg
with_pg
$ perl myapp.pl no_pg
no_pg
```

Also, mode is not intended to be limited to just production and development; in fact, Mojolicious intentionally doesn't care what mode you're operating in.  It does have some built-in behavior for the production mode, such as automatically changing the behavior of the App's log object, but you still have the ability to override that.  And documentation uses production and development to illustrate examples.  But, really, you can -- and should -- define whatever modes you want with whatever behavior you want.  I define a "test" mode for my App's tests to, for example, bypass an under route which requires authorization.  I also use a "cli" mode to simplify the not_found and exception pages so that I'm not blasted with screens of HTML when all I need to see is the 404 or exception message.

**Reply by augensalat** [reply]
Hi Stefan, thank you for the detailed description, but I still don't see how this can help.

Following your suggestion my application now basically looks like this:

```perl
package MyApp;
use Mojo::Base 'Mojolicious', -signatures;

sub startup ($self) {
    my $config = $self->plugin('NotYAMLConfig');

    $self->hook(before_command => sub ($command, $) {
        $self->log->debug("Command " . ref($command));
    });

    $self->log->debug('Minion');
    $self->plugin(Minion => {Pg => $config->{database}});
    ...
}
```

When I run `my_app version` I see this in the logs:

```
[2021-11-20 21:53:45.37763] [36] [debug] Minion
[2021-11-20 21:53:45.46128] [36] [debug] Command Mojolicious::Command::version
```

The `startup` method is run before the `before_command` hook.

And when I run the actual dev web application with `morbo` the `before_command` hook is not reached at all.

So I still don't see how I would skip Minion in (certain) commands and activate it for the web app mode.

**Reply by s1037989** [reply]
Hmm, yes, that's quite a pickle! Thanks for the clear report about how that all worked out.

Ok, what if you move the call to load the Minion plugin into a `before_server_start` hook? That should get called on web apps but not by commands. Therefore, commands won't load the plugin, but the web app will!

"Useful for reconfiguring application servers dynamically" ... That's exactly what you're looking for right? With that proposed configuration, your app by default doesn't load Minion, but instead reconfigures the application server dynamically by loading Minion `before_server_start`.

However, how will commands that need Minion (such as the minion commands) be able to function because the Minion plugin won't get loaded? Well, go back to the `before_command` hook and specify the commands that should load the Minion plugin. You can choose to implement that where it'll load Minion by default except for these commands, or only load Minion on these commands.

I'd hate to have to suggest a bunch of if statements or a regular expression. Perhaps the cleanest way would be to add an attribute to commands that shouldn't load Minion. `has 'skip_minion'`. And when that's set, `before_command` will skip loading Minion for that command, otherwise it loads it. (I assume that `before_command` would enable reaching into the command for an attribute.)

By moving Minion plugin load to both hooks, it sounds like it will get loaded only once, depending on application invocation. In other words, invocation is either a command or a server, but not both. It could become a problem, I suppose, if you have a command that needs Minion and also loads a server. It's not really a problem, I don't think, but it does register the plugin both times in that rare situation.

A final scenario is an innovation where you neither run a command nor start a server. I think that's technically possible, and what happens is you get an application usage response. But, what if you intended to run your app in some other way, similar to morbo. You wouldn't need to call a command, and you wouldn't need to start a server. In this case, the Minion plugin wouldn't load. Instead, you'd have to load it directly in order to use it.

Hopefully you can figure out something that works, and relatively elegantly. Hoping to be able to see what you come with, or what the next pickle is!

**Reply by augensalat** [reply]
Hi Stefan,

so I changed MyApp to this (basically):

```perl
package MyApp;
use Mojo::Base 'Mojolicious', -signatures;

sub startup ($self) {
    my $config = $self->plugin('NotYAMLConfig');

    $self->hook(before_command => sub ($command, $) {
        $self->log->debug("Command " . ref($command));
        $self->plugin(Minion => {Pg => $config->{database}}) if ref($command) =~ /^Minion::Command/;
    });

    $self->hook(before_server_start => sub ($server, $) {
        $self->log->debug("Server " . ref($server));
        $self->plugin(Minion => {Pg => $config->{database}}
    }
    ...
}

```

This looks promising.
Starting the web application with morbo works.
`my_app version` works now as well.

But...

```
$ my_app minion job
Unknown command "minion", maybe you need to install it?
```

Yes, Minion is not there (even when I remove the condition `if ref($command) =~ /^Minion::Command/` just to be sure):

```
my_app
...
Commands:
 cgi       Start application with CGI
 cpanify   Upload distribution to CPAN
 daemon    Start application with HTTP and WebSocket server
 eval      Run code against application
 generate  Generate files and directories from templates
 get       Perform HTTP request
 inflate   Inflate embedded files to real files
 prefork   Start application with pre-forking HTTP and WebSocket server
 psgi      Start application with PSGI
 routes    Show available routes
 version   Show versions of available modules

See 'APPLICATION help COMMAND' for more information on a specific command.
```

It seems loading the `Minion::Command`s via the `Mojolicious::Plugin::Minion` in the `before_command` hook is simply too late, IOW the check for the available commands happens earlier. I assume it is this line, which must have been executed earlier:
https://metacpan.org/dist/Minion/source/lib/Mojolicious/Plugin/Minion.pm#L8

So thanks buddy, but again no luck. :-/

Your last point about the final scenario is too abstract for me, but I have the feeling that this is currently not my biggest concern.

**Reply by s1037989** [reply]
Perhaps call [L8](https://metacpan.org/dist/Minion/source/lib/Mojolicious/Plugin/Minion.pm#L8) in your own app's startup, not a hook?  Will that solve all the problems?  If so, could you create your own `MinionWrapper` plugin (or Minion-alternative plugin) that will load unconditionally, push `Minion::Command` onto the stack, and conditionally load Minion?

```perl
package MyApp;
use Mojo::Base 'Mojolicious', -signatures;

sub startup ($self) {
    my $config = $self->plugin('NotYAMLConfig');
    $self->plugin('MyMinion' => {Pg => $config->{database}});
}

package Mojolicious::Plugin::MyMinion;
use Mojo::Base 'Mojolicious::Plugin', -signatures;

sub register {
    my ($self, $app, $conf) = @_;
    push @{$app->commands->namespaces}, 'Minion::Command';
    $self->hook(before_command => sub ($command, $) {
        $self->log->debug("Command " . ref($command));
        $self->_plugin_minion($app, $conf) if ref($command) =~ /^Minion::Command/;
    });
    $self->hook(before_server_start => sub ($server, $) {
        $self->log->debug("Server " . ref($server));
        $self->_plugin_minion($app, $conf);
    });
}
sub _plugin_minion ($self, $app, $conf) {
    my $minion = Minion->new(%$conf)->app($app);
    $app->helper(minion => sub {$minion});
}
```

**Reply by s1037989** [reply]
Here's an interesting technique, but I reckon it'll still depend on conditional activation, likely a configuration.  Use a Dummy backend for Minion:

```perl
package Minion::Backend::Dummy;
use Mojo::Base 'Minion::Backend';
 
sub auto_retry_job    { 0 }
sub broadcast         { 0 }
sub dequeue           { {} }
sub enqueue           { undef }
sub fail_job          { 0 }
sub finish_job        { 0 }
sub history           { {} }
sub list_jobs         { {} }
sub list_locks        { {} }
sub list_workers      { {} }
sub lock              { 0 }
sub new               { shift->SUPER::new }
sub note              { 0 }
sub receive           { [] }
sub register_worker   { undef }
sub remove_job        { 0 }
sub repair            { }
sub reset             { }
sub retry_job         { 0 }
sub stats             { {} }
sub unlock            { 0 }
sub unregister_worker { }
```

Use it like so:
```
$ perl -Ilib2 -Mojo -E 'plugin "Minion" => {Dummy => 1}; app->start' minion job
$ perl -Ilib2 -Mojo -E 'plugin "Minion" => {Dummy => 1}; app->start' minion 
Usage: APPLICATION minion COMMAND [OPTIONS]

Commands:
 job     Manage Minion jobs
 worker  Start Minion worker

See 'APPLICATION minion help COMMAND' for more information on a specific command.
```

**Reply by augensalat** [reply]
Well, I want so solve a problem, but without rewriting large parts of the used CPAN module(s).
Fortunately the application is wrapped in a Docker image, where I am Boss. ;-)
So my own solution is to simply patch away the problematic three lines of `Minion::Backend::Pg` in the Dockerfile.

But since I still think that the version check in that module is not a good idea (or not implemented in the best way) I opened this issue: https://github.com/mojolicious/minion/issues/121
